### DISCUSSION TITLE
Enforce type hints on specific methods

### DISCUSSION THREAD

**Discussion by epenet** [home-assistant/architecture#712] [General] [answered]
I have used a custom script to add common type hints in the code, and I believe that automatic checks should be added to the CI in order to ensure those methods are always typed in future PRs.

Here are the methods that I would recommend enforcing as a starting point:

Inside `__init__.py`:
```python
def setup(hass: HomeAssistant, config: ConfigType) -> bool:
def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
def async_setup_entry(hass: HomeAssistant, config: ConfigEntry) -> bool:
def async_remove_entry(hass: HomeAssistant, config: ConfigEntry) -> None:
def async_unload_entry(hass: HomeAssistant, config: ConfigEntry) -> bool:
def async_migrate_entry(hass: HomeAssistant, config: ConfigEntry) -> bool:
```

Inside `{Platform.XXX}.py`:
```python
def setup_platform(hass: HomeAssistant, config: ConfigType, add_entities: AddEntitiesCallback, discovery_info: DiscoveryInfoType | None = None) -> None:
async def async_setup_platform(hass: HomeAssistant, config: ConfigType, async_add_entities: AddEntitiesCallback, discovery_info: DiscoveryInfoType | None = None) -> None:
async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback) -> None:
```

**Comment by bdraco** [comment]
Enforcing typing makes it much more difficult for first time contributors as the number of people that understand python typing is far less than the number who can potentially write integrations.

Strict enforcement of a typing requirement will have the side effect of moving more integration to HACS which is the opposite of the goal here (improving the code quality).
 
If we want to go this route, we should  automate adding the typing in some way that doesn't place additional burden on potential contributors.

**Reply by epenet** [reply]
Draft PR https://github.com/home-assistant/core/pull/64307 is an attempt to enforce the type hint via a pre-commit hook (like isort).
This would be automatic, and the user wouldn't have to think about it.

**Reply by epenet** [reply]
Draft PR home-assistant/core#64313 is an attempt to enforce the type hint via pylint.
I can tweak it in such a way that it would be ignored if the return value isn't typed, and none of the arguments are typed either.

**Comment by thecode** [comment]
I don't think we should enforce specific methods, we should either enforce full strict typing or no enforcement at all. If we would enforce something enforcing tests has better benefit on the stability of the project, but both would be to difficult for first time contributors. 

What we can do however is make sure the full codebase uses type hints and our demo and scaffold generate code with type hints as most users probably copy code from somewhere else. It is already very hard for new contributors to get into the project and we should think of how we make their entrance easier.

The state today is that we also have contributors which does not have a specific hardware or knowledge how to interact with the hardware but do have good python skills and usually they are used as a 2nd pass to go over code and add type hints.

**Reply by frenck** [reply]
I kinda agree with this.

**Reply by epenet** [reply, marked as answer]
This PR only applies if the method has some type hints in. If the method doesn't have any type hints (argument or return), then the plugin doesn't trigger: https://github.com/home-assistant/core/pull/64313
Then it stays optional, but we ensure that the hints are correct if the contributor has chosen to add them in.

**Reply by frenck** [reply]
When taking that approach, I'm in favor 👍

**Comment by Misiu** [comment]
Full strict typing would be awesome.
I've created a couple of integrations and it was pure pain at the beginning, because of missing typings. I had to search in other integrations for example of specific methods. Strict typing, in my opinion, would help, because autocomplete would fully work and at least you would know the type that the method should return. I still have problems with writing tests, but that's another story.

HACS is great and soon many integrations will move there (because of https://github.com/home-assistant/architecture/blob/master/adr/0019-GPIO.md). Home Assistant is adding typings, there are new rules for mypy and the code quality is improving very fast, but custom components are on their own.
When developing s custom component I've forked the core repo and first added my integration there (before moving it to a separate repo) to be able to use all the goodness HA repository has configured.

If we could (optionally) use the same rules and checks when developing custom components that would help keep up with the quality of the core integrations.
